core/docs/sources-asciidoc/src/main/asciidoc/Section-JDiameter-Configuration.adoc

[[_jdiameter_configuration]]
= Diameter Stack Configuration

The stack is initially configured by parsing an XML file.
The top level structure of the file is described below.
Further explanation of each child element, and the applicable attributes, is provided later in this section.

[source,xml]
----
<Configuration xmlns="http://www.jdiameter.org/jdiameter-server">

	<LocalPeer></LocalPeer>
	<Parameters></Parameters>
	<Network></Network>
	<Extensions></Extensions>

</Configuration>
----

[source,xml]
----
<LocalPeer>
	<URI value="aaa://localhost:3868"/>
	<IPAddresses>
		<IPAddress value="127.0.0.1"/>
	</IPAddresses>

	<Realm value="mobicents.org"/>
	<VendorID value="193"/>
	<ProductName value="jDiameter"/>
	<FirmwareRevision value="1"/>

	<OverloadMonitor>
		<Entry index="1" lowThreshold="0.5" highThreshold="0.6">
			<ApplicationID>
				<VendorId value="193"/>
				<AuthApplId value="0"/>
				<AcctApplId value="19302"/>
			</ApplicationID>
		</Entry>
	</OverloadMonitor>
	<Applications>
		<ApplicationID>
			<VendorId value="193"/>
			<AuthApplId value="0"/>
			<AcctApplId value="19302"/>
		</ApplicationID>
	</Applications>
</LocalPeer>
----

The <LocalPeer> element contains parameters that affect the local Diameter peer.
The available elements and attributes are listed for reference.

.<LocalPeer> Elements and Attributes
<URI>::
  Specifies the URI for the local peer.
  The URI has the following format: "aaa://FQDN:port".

<IPAddresses>::
  Contains one or more child <IPAddress> element, which contain a single, valid IP address for the local peer, stored in the [parameter]`value` attribute of the IPAddress.

<Realm>::
  Specifies the realm of the local peer, using the [parameter]`value` attribute.

<VendorID>::
  Specifies a numeric identifier that corresponds to the vendor ID allocated by IANA. 

<ProductName>::
  Specifies the name of the local peer product.

<FirmwareRevision>::
  Specifies the version of the firmware.

<OverloadMonitor>::
  Optional parent element containing child elements that specify settings relating to the Overload Monitor.

<Entry>::
  Supports <ApplicationID> child elements that specify the ID of the tracked application(s). It also supports the following properties: 

index
  Defines the index of this overload monitor, so priorities/orders can be specified.

lowThreshold
  The low threshold for activation of the overload monitor.

highThreshold
  The high threshold for activation of the overload monitor.

<ApplicationID>::
  Parent element containing child elements that specify information about the application.
  The child elements create a unique application identifier.
  The child elements are:

<VendorId>
Specifies the vendor ID for application definition. It supports a single property: "value".

<AuthAppId>
The Authentication Application ID for application definition. It supports a single property: "value".

<AcctAplId>
The Account Application ID for application definition. It supports a single property: "value".

<Applications>::
  Contains a child element <ApplicationID>, which defines the list of default supported applications.
  It is used for the server side, when the stack is configured to accept incoming calls and there is an empty list of preconfigured peers (server is configured to accept any connection).

[source,xml]
----
<Parameters>

	<AcceptUndefinedPeer value="true"/>
	<DuplicateProtection value="true"/>
    <DuplicateTimer value="240000"/>
    <DuplicateSize value="5000"/>
	<UseUriAsFqdn value="true"/> <!-- Needed for Ericsson SDK Emulator -->
	<QueueSize value="10000"/>
	<MessageTimeOut value="60000"/>
	<StopTimeOut value="10000"/>
	<CeaTimeOut value="10000"/>
	<IacTimeOut value="30000"/>
	<DwaTimeOut value="10000"/>
	<DpaTimeOut value="5000"/>
	<RecTimeOut value="10000"/>
	<SessionInactivityTimeOut value="600"/>

	<!-- Peer FSM Thread Count Configuration -->
	<PeerFSMThreadCount value="3" />

	<Concurrent>
		<Entity name="ThreadGroup" size="64"/>
		<Entity name="ProcessingMessageTimer" size="1"/>
		<Entity name="DuplicationMessageTimer" size="1"/>
		<Entity name="RedirectMessageTimer" size="1"/>
		<Entity name="PeerOverloadTimer" size="1"/>
		<Entity name="ConnectionTimer" size="1"/>
		<Entity name="StatisticTimer" size="1"/>
		<Entity name="ApplicationSession" size="16"/>
	</Concurrent>

</Parameters>
----

The <Parameters> element contains elements that specify parameters for the Diameter stack.
The available elements and attributes are listed for reference.
If not specified otherwise, each tag supports a single property - "value", which indicates the value of the tag.

.<Parameters> Elements and Attributes
<AcceptUndefinedPeer>::
  Specifies whether the stack will accept connections from undefined peers.
  The default value is `false`.

<DuplicateProtection>::
  Specifies whether duplicate message protection is enabled.
  The default value is `false`.

<DuplicateTimer>::
  Specifies the time each duplicate message is valid for (in extreme cases, it can live up to 2 * DuplicateTimer - 1 milliseconds). The default, minimum value is `240000` (4 minutes in milliseconds).

<DuplicateSize>::
  Specifies the number of requests stored for duplicate protection.
  The default value is `5000`.

<UseUriAsFqdn>::
  Determines whether the URI should be used as FQDN.
  If it is set to `true`, the stack expects the destination/origin host to be in the format of "aaa://isdn.domain.com:3868" rather than the normal "isdn.domain.com". The default value is `false`.

<QueueSize>::
  Determines how many tasks the peer state machine can have before rejecting the next task.
  This queue contains FSM events and messaging.

<MessageTimeOut>::
  Determines the timeout for messages other than protocol FSM messages.
  The delay is in milliseconds.

<StopTimeOut>::
  Determines how long the stack waits for all resources to stop.
  The delays are in milliseconds.

<CeaTimeOut>::
  Determines how long it takes for CER/CEA exchanges to timeout if there is no response.
  The delays are in milliseconds.

<IacTimeOut>::
  Determines how long the stack waits to retry the communication with a peer that has stopped answering DWR messages.
  The delay is in milliseconds.

<DwaTimeOut>::
  Determines how long it takes for a DWR/DWA exchange to timeout if there is no response.
  The delay is in milliseconds.

<DpaTimeOut>::
  Determines how long it takes for a DPR/DPA exchange to timeout if there is no response.
  The delay is in milliseconds.

<RecTimeOut>::
  Determines how long it takes for the reconnection procedure to timeout.
  The delay is in milliseconds.

<SessionInactivityTimeOut>::
  Determines how much time the persistence record should be kept if there is no request sent within a session.
  Irrelevant when session persistent routing is not enabled. The delay is in seconds.

<PeerFSMThreadCount>::
  Determines the number of threads for handling events in the Peer FSM.

<Concurrent />::
  Controls the thread pool sizes for different aspects of the stack.
  It supports multiple [parameter]`Entity` child elements. [parameter]`Entity` elements configure thread groups.
  These elements support the following properties:

name
Specifies the name of the entity.

size
Specifies the thread pool size of the entity.

The default supported entities are:

ThreadGroup
Determines the maximum thread count in other entities.

ProcessingMessageTimer
Determines the thread count for message processing tasks.

DuplicationMessageTimer
Specifies the thread pool for identifying duplicate messages.

RedirectMessageTimer
Specifies the thread pool for redirecting messages that do not need any further processing.

PeerOverloadTimer
Determines the thread pool for managing the overload monitor.

ConnectionTimer
Determines the thread pool for managing tasks regarding peer connection FSM.

StatisticTimer
Determines the thread pool for statistic gathering tasks.

ApplicationSession
Determines the thread pool for managing the invocation of application session FSMs, which will invoke listeners.

[source,xml]
----
<Network>

	<Peers>
		<!-- This peer is a server, if it's a client attempt_connect should be set to false -->
		<Peer name="aaa://127.0.0.1:3868" attempt_connect="true" rating="1"/>
	</Peers>

	<Realms>
		<Realm name="mobicents.org" peers="127.0.0.1" local_action="LOCAL" dynamic="false" exp_time="1">
			<ApplicationID>
				<VendorId value="193"/>
				<AuthApplId value="0"/>
				<AcctApplId value="19302"/>
			</ApplicationID>
		</Realm>
	</Realms>

</Network>
----

The <Network> element contains elements that specify parameters for external peers.
The available elements and attributes are listed for reference.

.<Network> Elements and Attributes
<Peers>::
  Parent element containing the child element <Peer>, which specifies external peers and the way they connect.
  <Peer> specifies the name of external peers, whether they should be treated as a server or client, and what rating the peer has externally.

<Peer> supports the following properties:

name
Specifies the name of the peer in the form of a URI. The structure is "aaa://[fqdn|ip]:port" (for example, "aaa://192.168.1.1:3868").

attempt_connect
Determines if the stack should try to connect to this peer. This property accepts boolean values.

rating
Specifies the rating of this peer in order to achieve peer priorities/sorting.

<Realms>::
  Parent element containing the child element <Realm>, which specifies all realms that connect into the Diameter network.
  <Realm> contains attributes and elements that describe different realms configured for the Core.
  It supports <ApplicationID> child elements, which define the applications supported.

<Realm> supports the following parameters:

peers
Comma separated list of peers. Each peer is represented by an IP Address or FQDN.

local_action
Determines the action the Local Peer will play on the specified realm: Act as a LOCAL peer.

dynamic
Specifies if this realm is dynamic. That is, peers that connect to peers with this realm name will be added to the realm peer list if not present already.

exp_time
The time before a peer belonging to this realm is removed if no connection is available.


Below is an example configuration file for a server supporting the CCA, Sh and Ro Applications:

[source,xml]
----
<?xml version="1.0"?>
<Configuration xmlns="http://www.jdiameter.org/jdiameter-server">

	<LocalPeer>
		<URI value="aaa://127.0.0.1:3868" />
		<Realm value="mobicents.org" />
		<VendorID value="193" />
		<ProductName value="jDiameter" />
		<FirmwareRevision value="1" />
		<OverloadMonitor>
			<Entry index="1" lowThreshold="0.5" highThreshold="0.6">
				<ApplicationID>
					<VendorId value="193" />
					<AuthApplId value="0" />
					<AcctApplId value="19302" />
				</ApplicationID>
			</Entry>
		</OverloadMonitor>
	</LocalPeer>

	<Parameters>
		<AcceptUndefinedPeer value="true" />
		<DuplicateProtection value="true" />
    <DuplicateTimer value="240000" />
    <DuplicateSize value="5000" />
		<UseUriAsFqdn value="false" /> <!-- Needed for Ericsson Emulator (set to true) -->
		<QueueSize value="10000" />
		<MessageTimeOut value="60000" />
		<StopTimeOut value="10000" />
		<CeaTimeOut value="10000" />
		<IacTimeOut value="30000" />
		<DwaTimeOut value="10000" />
		<DpaTimeOut value="5000" />
		<RecTimeOut value="10000" />
		<SessionInactivityTimeOut value="600"/>

		<PeerFSMThreadCount value="3" />

		<Concurrent>
			 <Entity name="ThreadGroup" size="64"/>
			 <Entity name="ProcessingMessageTimer" size="1"/>
			 <Entity name="DuplicationMessageTimer" size="1"/>
			 <Entity name="RedirectMessageTimer" size="1"/>
			 <Entity name="PeerOverloadTimer" size="1"/>
			 <Entity name="ConnectionTimer" size="1"/>
			 <Entity name="StatisticTimer" size="1"/>
			 <Entity name="ApplicationSession" size="16"/>
		</Concurrent>
	</Parameters>

	<Network>
		<Peers>
			<Peer name="aaa://127.0.0.1:1218" attempt_connect="false" rating="1" />
		</Peers>
		<Realms>
			<!-- CCA -->
			<Realm name="mobicents.org" peers="127.0.0.1" local_action="LOCAL" 
				dynamic="false" exp_time="1">
				<ApplicationID>
					<VendorId value="0" />
					<AuthApplId value="4" />
					<AcctApplId value="0" />
				</ApplicationID>
			</Realm>
			
			<!-- Sh -->
			<Realm name="mobicents.org" peers="127.0.0.1" local_action="LOCAL" 
				dynamic="false" exp_time="1">
				<ApplicationID>
					<VendorId value="10415" />
					<AuthApplId value="16777217" />
					<AcctApplId value="0" />
				</ApplicationID>
			</Realm>

			<!-- Ro -->
			<Realm name="mobicents.org" peers="127.0.0.1" local_action="LOCAL" 
				dynamic="false" exp_time="1">
				<ApplicationID>
					<VendorId value="10415" />
					<AuthApplId value="4" />
					<AcctApplId value="0" />
				</ApplicationID>
			</Realm>
		</Realms>
	</Network>

	<Extensions />

</Configuration>
----

[[_jdiameter_cluster_configuration]]
== Cluster configuration

The following list defines the requirements for enabling stack cluster mode

* Add the following entries to the `Parameters` section of [path]_jdiameter-config.xml_: 
+
[source,xml]
----

				
<SessionDatasource>org.mobicents.diameter.impl. ha.data.ReplicatedData</SessionDatasource>
<TimerFacility>org.mobicents.diameter.impl.ha. timer.ReplicatedTimerFacilityImpl</TimerFacility>
----

* A proper `JBoss Cache` configuration file: [path]_jdiameter-jbc.xml_ (located in the [path]_config_ directory).
+
The following content is sufficient for the JBoss Cache configuration file: 
+
[source,xml]
----

				
<?xml version="1.0" encoding="UTF-8"?>

<jbosscache xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns="urn:jboss:jbosscache-core:config:3.0">

	<locking isolationLevel="REPEATABLE_READ"
		lockParentForChildInsertRemove="false" lockAcquisitionTimeout="20000"
		nodeLockingScheme="mvcc" writeSkewCheck="false" concurrencyLevel="500" />

	<jmxStatistics enabled="false" />

	<startup regionsInactiveOnStartup="false" />
	<shutdown hookBehavior="DEFAULT" />
	<listeners asyncPoolSize="1" asyncQueueSize="100000" />

	<invocationBatching enabled="false" />

	<serialization objectInputStreamPoolSize="12"
		objectOutputStreamPoolSize="14" version="3.0.0"
		marshallerClass="org.jboss.cache.marshall.CacheMarshaller300"
		useLazyDeserialization="false" useRegionBasedMarshalling="false" />

	<clustering mode="replication" clusterName="DiameterCluster">

		<async useReplQueue="true" replQueueInterval="1000"
			replQueueMaxElements="500" serializationExecutorPoolSize="20"
			serializationExecutorQueueSize="5000000"/>

		<jgroupsConfig>
			<UDP
				mcast_addr="${jgroups.udp.mcast_addr:228.10.10.10}"
				mcast_port="${jgroups.udp.mcast_port:18811}"
				discard_incompatible_packets="true"
				max_bundle_size="60000"
				max_bundle_timeout="30"
				ip_ttl="${jgroups.udp.ip_ttl:2}"
				enable_bundling="true"
				thread_pool.enabled="true"
				thread_pool.min_threads="1"
				thread_pool.max_threads="25"
				thread_pool.keep_alive_time="5000"
				thread_pool.queue_enabled="false"
				thread_pool.queue_max_size="100"
				thread_pool.rejection_policy="Run"
				oob_thread_pool.enabled="true"
				oob_thread_pool.min_threads="1"
				oob_thread_pool.max_threads="8"
				oob_thread_pool.keep_alive_time="5000"
				oob_thread_pool.queue_enabled="false"
				oob_thread_pool.queue_max_size="100"
				oob_thread_pool.rejection_policy="Run"/>

			<PING timeout="2000"
				num_initial_members="3"/>
			<MERGE2 max_interval="30000"
				min_interval="10000"/>
			<FD_SOCK/>
			<FD timeout="10000" max_tries="5" />
			<VERIFY_SUSPECT timeout="1500"  />
			<BARRIER />
			<pbcast.NAKACK
				use_mcast_xmit="false" gc_lag="0"
				retransmit_timeout="300,600,1200,2400,4800"
				discard_delivered_msgs="true"/>
			<UNICAST timeout="300,600,1200,2400,3600"/>
			<pbcast.STABLE stability_delay="1000" desired_avg_gossip="50000"
				max_bytes="400000"/>
			<VIEW_SYNC avg_send_interval="60000"   />
			<pbcast.GMS print_local_addr="true" join_timeout="3000"
				view_bundling="true"/>
			<FC max_credits="20000000"
				min_threshold="0.10"/>
			<FRAG2 frag_size="60000"  />
			<pbcast.STATE_TRANSFER  />
		</jgroupsConfig>
	</clustering>

</jbosscache>
----

[[_jdiameter_failover_configuration]]
== Failover configuration

Apart from a default routing scheme, which does not require any additional configuration,
there is an option of activating failure aware routing that extends capabilities of basic
router with extra features related to failure detection, peer priority handling and load
balancing. Rating of a particular peer is taken into consideration when deciding about
an order of peers usage in case of failure detection. The highest rating peers are used first,
then lower priorities peers next, etc. If several peers are marked with the same rating,
load balancing algorithm is executed among them. In case of all higher priority peers failure,
lower priority peers are considered. Afterwards, in case any higher priority peer becomes
available again and session persistence is enabled as well, only new sessions requests are 
targeted again to higher priority peers, i.e. currently handled session stays assigned to 
the peer selected beforehand.

In order to enable a/m extended routing feature, the following entry has to be added to the `Extensions`
section of [path]_jdiameter-config.xml_:

[source,xml]
----
<RouterEngine>org.jdiameter.server.impl.FailureAwareRouter</RouterEngine>
----

Additionally, along with an extended routing policy, it is also highly advised to enable session
persistence as well. Otherwise, routing decisions will be made for every single request within
a particular session what may eventually result in multiple undesirable destination peer reselections.

[[_jdiameter_session_persistence_configuration]]
== Session persistence

Session persistence enforces sticky sessions that map a single diameter session to a single peer
which had been selected to process such a session. Session persistence record is created after
a peer had answered the first (initial) request for that session. Furthermore, it can be updated
in the event of peer reselection by failover algorithm. Finally, it is removed when session is finished
normally, an error indication answer is received or session inactivity timeout expires. Replication of
session persistence records is not supported.

The following list defines the requirements for enabling session persistence:

* Add the following entry to the `Parameters` section of [path]_jdiameter-config.xml_: 
+
[source,xml]
----
<SessionDatasource>org.jdiameter.common.impl.data.RoutingAwareDataSource</SessionDatasource>
----

* Customize the value of `SessionInactivityTimeOut` in the `Extensions` section of [path]_jdiameter-config.xml_

If enabled, session persistence feature supports two types of applications, i.e. CCA (defined in 
http://tools.ietf.org/html/rfc4006[RFC 4006]) and Ro (defined in http://www.3gpp.org/DynaReport/32240.htm[3GPP TS 32.240]
and http://www.3gpp.org/DynaReport/32299.htm[3GPP TS 32.299]) by virtue of their session based specificity.


core/jdiameter/impl/src/main/java/org/jdiameter/client/impl/helpers/Parameters.java
/*
 * JBoss, Home of Professional Open Source
 * Copyright 2006, Red Hat, Inc. and individual contributors
 * by the @authors tag. See the copyright.txt in the distribution for a
 * full listing of individual contributors.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

package org.jdiameter.client.impl.helpers;

import java.util.ArrayList;

/**
 * This enumeration defined all parameters of diameter stack implementation
 * 
 * @author erick.svenson@yahoo.com
 * @author <a href="mailto:brainslog@gmail.com"> Alexandre Mendonca </a>
 * @author <a href="mailto:baranowb@gmail.com"> Bartosz Baranowski </a>
 */
public class Parameters extends Ordinal {

  private static final long serialVersionUID = 1L;

  protected static int index;

  private static ArrayList<Parameters> value = new ArrayList<Parameters>();

  /**
   * Class name of IOC  property
   */
  public static final Parameters Assembler = new Parameters("Assembler", String.class, "org.jdiameter.client.impl.helpers.AssemblerImpl");

  /**
   * Local peer URI property
   */
  public static final Parameters OwnDiameterURI = new Parameters("OwnDiameterURI", String.class, "aaa://localhost:3868");
  /**
   * Local peer ip address property
   */
  public static final Parameters OwnIPAddress = new Parameters("OwnIPAddress", String.class, "");
  /**
   * Local peer realm name property
   */
  public static final Parameters OwnRealm = new Parameters("OwnRealm", String.class, "local");
  /**
   * Local peer vendor id  property
   */
  public static final Parameters OwnVendorID = new Parameters("OwnVendorID", Long.class, 0L);
  /**
   * Local peer stack product name property
   */
  public static final Parameters OwnProductName = new Parameters("OwnProductName", String.class, "jDiameter");
  /**
   * Local peer stack firmware version property
   */
  public static final Parameters OwnFirmwareRevision = new Parameters("OwnFirmwareRevision", Long.class, 0L);

  /**
   * Task executor task queue size  property 
   */
  public static final Parameters QueueSize = new Parameters("QueueSize", Integer.class,   10000);

  /**
   * Message time out  property
   */
  public static final Parameters MessageTimeOut = new Parameters("MessageTimeOut", Long.class, 60000L);
  /**
   * Stop stack time out  property
   */
  public static final Parameters StopTimeOut = new Parameters("StopTimeOut", Long.class, 10000L);

  /**
   * CEA command time out property
   */
  public static final Parameters CeaTimeOut = new Parameters("CeaTimeOut", Long.class, 10000L);
  /**
   * Peer inactive time out property
   */
  public static final Parameters IacTimeOut = new Parameters("IacTimeOut", Long.class, 20000L);
  /**
   * DWA command time out property
   */
  public static final Parameters DwaTimeOut = new Parameters("DwaTimeOut", Long.class, 10000L);
  /**
   * DPA command time out property
   */
  public static final Parameters DpaTimeOut = new Parameters("DpaTimeOut", Long.class, 5000L);
  /**
   * Reconnect time out property
   */
  public static final Parameters RecTimeOut = new Parameters("RecTimeOut", Long.class, 10000L);

  /**
   * Peer FSM Thread Count property
   */
  public static final Parameters PeerFSMThreadCount = new Parameters("PeerFSMThreadCount", Integer.class, 3);

  /**
   * Orig_host avp set as URI into CER message
   */
  public static final Parameters UseUriAsFqdn = new Parameters("UseUriAsFqdn", Boolean.class, false);    

  /**
   * Peer name property
   */
  public static final Parameters PeerName = new Parameters("PeerName", String.class, "");

  /**
   * Peer ip property
   */
  public static final Parameters PeerIp = new Parameters("PeerIp", String.class, "");

  /**
   * Peer local peer port range (format: 1345-1346) property
   */
  public static final Parameters PeerLocalPortRange = new Parameters("PeerLocalPortRange", String.class, "");

  /**
   * Peer rating property
   */
  public static final Parameters PeerRating = new Parameters("PeerRating", Integer.class, 0);
  /**
   *  Peer ptoperty
   */
  public static final Parameters Peer = new Parameters("Peer", Object.class);

  /**
   * Real entry property
   */
  public static final Parameters RealmEntry = new Parameters("RealmEntry", String.class, "");
  /**
   * Agent element
   */
  public static final Parameters Agent = new Parameters("Agent", Object.class);
  /**
   * Properties element
   */
  public static final Parameters Properties = new Parameters("Properties", Object.class);
  /**
   * Properties element
   */
  public static final Parameters Property = new Parameters("Property ", Object.class);
  /**
   * Property name
   */
  public static final Parameters PropertyName = new Parameters("name", String.class, "");
  /**
   * Property value
   */
  public static final Parameters PropertyValue = new Parameters("value", String.class, "");
  /**
   * Realm property
   */
  public static final Parameters Realm = new Parameters("Realm", Object.class);

  /**
   * Vendor id property
   */
  public static final Parameters VendorId = new Parameters("VendorId", Long.class);
  /**
   * Authentication application id property
   */
  public static final Parameters AuthApplId = new Parameters("AuthApplId", Long.class);
  /**
   * Accounting application id property
   */
  public static final Parameters AcctApplId = new Parameters("AcctApplId", Long.class);

  /**
   * Application Id property
   */
  public static final Parameters ApplicationId = new Parameters("ApplicationId", Object.class);

  /**
   * Extension point property
   */
  public static final Parameters Extensions = new Parameters("Extensions", Object.class);
  /**
   * Extension point name property
   */
  public static final Parameters ExtensionName = new Parameters("ExtensionName", String.class);

  /**
   * Peer list property
   */
  public static final Parameters PeerTable = new Parameters("PeerTable", Object.class);

  /**
   * Realm list property
   */
  public static final Parameters RealmTable = new Parameters("RealmTable", Object.class);

  /**
   * Security list property
   */
  public static final Parameters Security = new Parameters("Security", Object.class);

  /**
   * Security entry
   */
  public static final Parameters SecurityData = new Parameters("SecurityData", Object.class);

  /**
   * Security data name
   */
  public static final Parameters SDName = new Parameters("SDName", String.class);    

  /**
   * Security protocol
   */
  public static final Parameters SDProtocol = new Parameters("SDProtocol", String.class, "TLS");

  /**
   * Security session creation flag
   */
  public static final Parameters SDEnableSessionCreation = new Parameters("SDEnableSessionCreation", Boolean.class, false);

  /**
   * Security client mode flag
   */
  public static final Parameters SDUseClientMode= new Parameters("SDUseClientMode", Boolean.class, false);

  /**
   * Cipher suites separated by ', '
   */
  public static final Parameters CipherSuites = new Parameters("CipherSuites", String.class);

  /**
   * Key data
   */
  public static final Parameters KeyData = new Parameters("KeyData", String.class);

  /**
   * Key manager
   */
  public static final Parameters KDManager = new Parameters("KDManager", String.class);

  /**
   * Key store
   */
  public static final Parameters KDStore = new Parameters("KDStore", String.class);

  /**
   * Key file
   */
  public static final Parameters KDFile = new Parameters("KDFile", String.class);

  /**
   * Key password
   */
  public static final Parameters KDPwd = new Parameters("KDPwd", String.class);


  /**
   * Trust data
   */
  public static final Parameters TrustData = new Parameters("TrustData", String.class);    

  /**
   * Key manager
   */
  public static final Parameters TDManager = new Parameters("TDManager", String.class);

  /**
   * Key store
   */
  public static final Parameters TDStore = new Parameters("TDStore", String.class);

  /**
   * Key file
   */
  public static final Parameters TDFile = new Parameters("TDFile", String.class);

  /**
   * Key password
   */
  public static final Parameters TDPwd = new Parameters("TDPwd", String.class);

  /**
   * Reference to security information
   */
  public static final Parameters SecurityRef = new Parameters("SecurityRef", String.class);

  /**
   * XML entry for thread pool
   */
  public static final Parameters ThreadPool = new Parameters("ThreadPool", Object.class);
  /**
   * Thread pool max size
   */
  public static final Parameters ThreadPoolSize = new Parameters("ThreadPoolSize", Integer.class, 10);
  /**
   * Thread pool max size
   */
  public static final Parameters ThreadPoolPriority = new Parameters("ThreadPoolPriority", Integer.class, 5);

  /**
   * Statistic logger properties
   */
  public static final Parameters Statistics = new Parameters("Statistics", Object.class);

  /**
   * Statistic logger start pause
   */
  public static final Parameters StatisticsLoggerPause = new Parameters("StatisticsLoggerPause", Long.class, 30000L);

  /**
   * Statistic logger delay between save statistic information
   */
  public static final Parameters StatisticsLoggerDelay = new Parameters("StatisticsLoggerDelay", Long.class, 30000L);

  /**
   * Statistic flag controlling if statistics are on/off - ie. timer tasks to display are created. 
   */
  public static final Parameters StatisticsEnabled = new Parameters("StatisticsEnabled", Boolean.class, false);

  /**
   * List of statistics names which should be enabled. 
   */
  public static final Parameters StatisticsActiveList = new Parameters("StatisticsActiveList", String.class, false);

  /**
   * Concurrent configuration root point
   */
  public static final Parameters Concurrent = new Parameters("Concurrent", Object.class);

  /**
   * Concurrent entity name
   */
  public static final Parameters ConcurrentEntityName = new Parameters("ConcurrentEntityName", String.class, "Empty");

  /**
   * Concurrent thread group name
   */
  public static final Parameters ConcurrentEntityDescription = new Parameters("ConcurrentEntityDescription", String.class, "ThreadPool");

  /**
   * Concurrent thread group size
   */
  public static final Parameters ConcurrentEntityPoolSize = new Parameters("ConcurrentEntityPoolSize", Integer.class, 4);

  /**
   * Dictionary root
   */
  public static final Parameters Dictionary = new Parameters("Dictionary", Object.class);

  /**
   * Dictionary Class name
   */
  public static final Parameters DictionaryClass = new Parameters("DictionaryClass", String.class, "org.jdiameter.common.impl.validation.DictionaryImpl");

  /**
   * Dictionary Validation enabled
   */
  public static final Parameters DictionaryEnabled = new Parameters("DictionaryEnabled", Boolean.class, false);

  /**
   * Dictionary Send Level Validation
   */
  public static final Parameters DictionarySendLevel = new Parameters("DictionarySendLevel", String.class, "MESSAGE");

  /**
   * Dictionary Receive Level Validation
   */
  public static final Parameters DictionaryReceiveLevel = new Parameters("DictionaryReceiveLevel", String.class, "OFF");

  /**
   * Maximum session inactivity time specified in seconds which defines how much time
   * the persistence record should be kept if there is no request sent within a session.
   * Irrelevant when session persistent routing is not enabled.
   */
  public static final Parameters SessionInactivityTimeOut = new Parameters("SessionInactivityTimeOut", Integer.class, 600);

  /**
   * Return all parameters as iterator
   * 
   * @return all parameters as iterator
   */
  public static Iterable<Parameters> values() {
    return value;
  }

  private Class type;
  private Object defValue;

  protected Parameters(String name, Class type) {
    this.name = name;
    this.type = type;
    ordinal = index++;
    value.add(this);
  }

  protected Parameters(String name, Class type, Object defValue) {
    this.name = name;
    this.type = type;
    this.defValue = defValue;
    ordinal = index++;
    value.add(this);
  }

  /**
   * Return default value of property
   * 
   * @return default value of property
   */
  public Object defValue() {
    return defValue;
  }

  /**
   * Return type of property
   * 
   * @return type of property
   */
  public Class type() {
    return type;
  }
}
